geopandas/tools/make_grid.py

```diff
@@ -0,0 +1,387 @@
+            )
+            if not flat_topped:
+                hex_centers_np = hex_centers_np[:, ::-1]
+
+            output_grid = points_from_xy(
+                hex_centers_np[:, 0] + grid_origin_x,
+                hex_centers_np[:, 1] + grid_origin_y,
+            )
+
+        elif what == "corners":
+            # The inverted center mask is the corner mask. Now consider all corners
+            mask_corners = np.invert(mask_center)
+
+            hex_corners_np = np.array([xv[mask_corners], yv[mask_corners]]).T.reshape(
+                -1, 2
+            )
+
+            if not flat_topped:
+                hex_corners_np = hex_corners_np[:, ::-1]
+
+            output_grid = points_from_xy(
+                hex_corners_np[:, 0] + grid_origin_x,
+                hex_corners_np[:, 1] + grid_origin_y,
+            )
+
+        elif what == "polygons":
+            hex_coords_a = _hex_polygon_corners(xv, yv, (0, 1))
+            hex_coords_b = _hex_polygon_corners(xv, yv, (2, 0))
+
+            hex_coords = np.concatenate((hex_coords_a, hex_coords_b), axis=0)
+
+            if not flat_topped:
+                hex_coords = hex_coords[:, :, ::-1]
+
+            hex_coords[:, :, 0] += grid_origin_x
+            hex_coords[:, :, 1] += grid_origin_y
+
+            hex_polygons = [Polygon(hex_set) for hex_set in hex_coords]
+
+            hex_polygons_np = np.array(hex_polygons)
+            output_grid = from_shapely(hex_polygons_np)
+
+        else:
+            raise ValueError("Invalid value for parameter `what`")
+    else:
+        raise ValueError("Invalid value for parameter `cell_type`")
+
+    output_grid = GeoSeries(output_grid)
+
+    if isinstance(input_geometry, (GeoDataFrame, GeoSeries)):
+        if input_geometry.crs is not None:
+            output_grid = output_grid.set_crs(input_geometry.crs)
+
+    if intersect:
+        if isinstance(input_geometry, (GeoDataFrame, GeoSeries)):
+            return output_grid[
+                output_grid.intersects(input_geometry.unary_union)
+            ].reset_index(drop=True)
+        else:
+            return output_grid[output_grid.intersects(input_geometry)].reset_index(
+                drop=True
+            )
+    else:
+        return output_grid
+
+
+def _hex_polygon_corners(xv, yv, index_0=(0, 0)):
+    """Helper function that groups hexagon corners that belong to the same grid cell
+
+    Parameters
+    ------------
+    xv : np.array
+        meshgrid containing x-values of all centers and corners of the hexgon grid
+    yv : np.array
+        meshgrid containing y-values of all centers and corners of the hexgon grid
+    index_0 : tuple
+        indicates index of the meshgrid, where polygon corner grouping starts
+
+    Returns
+    -------
+    np.array
+        The array has the shape (n, 6, 2) where n is the number of grid cells,
+        "6" corresponds to the hexagon corners and "2" to the corresponding
+        x,y coordinates.
+
+    """
+
+    # Input checks
```

Comment:
Do these errors happen? Since we're completely in control of input for this functions, we may not need those checks?